chips/nrf52/src/usbd.rs

```diff
@@ -0,0 +1,574 @@
+    /// Enable USB
+    /// - Address: 0x500 - 0x504
+    enable: ReadWrite<u32, Usb::Register>,
+    /// Control of the USB pull-up
+    /// - Address: 0x504 - 0x508
+    usbpullup: ReadWrite<u32, UsbPullup::Register>,
+    /// State at which the DPDMDRIVE task will force D+ and D-.
+    /// The DPDMNODRIVE task reverts the control of the lines
+    /// to MAC IP (no forcing).
+    /// - Address: 0x508 - 0x50C
+    dpdmvalue: ReadWrite<u32, DpDmValue::Register>,
+    /// Data toggle control and status
+    /// - Address: 0x50C - 0x510
+    dtoggle: ReadWrite<u32, Toggle::Register>,
+    /// Endpoint IN enable
+    /// - Address: 0x510 - 0x514
+    epinen: ReadWrite<u32, EndpointEnable::Register>,
+    /// Endpoint OUT enable
+    /// - Address: 0x514 - 0x518
+    epouten: ReadWrite<u32, EndpointEnable::Register>,
+    /// STALL endpoints
+    /// - Address: 0x518 - 0x51C
+    epstall: WriteOnly<u32, EndpointStall::Register>,
+    /// Controls the split of ISO buffers
+    /// - Address: 0x51C - 0x520
+    isosplit: ReadWrite<u32, IsoSplit::Register>,
+    /// Returns the current value of the start of frame counter
+    /// - Address: 0x520 - 0x524
+    framecntr: ReadOnly<u32, FrameCounter::Register>,
+    _reserved11: [u32; 2],
+    /// Controls USBD peripheral low power mode during USB suspend
+    /// - Address: 0x52C - 0x530
+    lowpower: ReadWrite<u32, LowPower::Register>,
+    /// Controls the response of the ISO IN endpoint to an IN token
+    /// when no data is ready to be sent
+    /// - Address: 0x530 - 0x534
+    isoinconfig: ReadWrite<u32, IsoInConfig::Register>,
+    _reserved12: [u32; 51],
+    /// - Address: 0x600 - 0x6A0
+    epin: [detail::EndpointRegisters; NUM_ENDPOINTS],
+    /// - Address: 0x6A0 - 0x6B4
+    isoin: detail::EndpointRegisters,
+    _reserved13: [u32; 19],
+    /// - Address: 0x700 - 0x7A0
+    epout: [detail::EndpointRegisters; NUM_ENDPOINTS],
+    /// - Address: 0x7A0 - 0x7B4
+    isoout: detail::EndpointRegisters,
+    _reserved14: [u32; 19],
+    /// Errata 166 related register (ISO double buffering not functional)
+    /// - Address: 0x800 - 0x804
+    errata166_1: WriteOnly<u32>,
+    /// Errata 166 related register (ISO double buffering not functional)
+    /// - Address: 0x804 - 0x808
+    errata166_2: WriteOnly<u32>,
+    _reserved15: [u32; 261],
+    /// Errata 199 related register (USBD cannot receive tasks during DMA)
+    /// - Address: 0xC1C - 0xC20
+    errata199: WriteOnly<u32>,
+}
+
+mod detail {
+    use super::{Amount, Count};
+    use kernel::common::cells::VolatileCell;
+    use kernel::common::registers::{ReadOnly, ReadWrite};
+
+    #[repr(C)]
+    pub struct EndpointRegisters {
+        ptr: VolatileCell<*const u8>,
+        maxcnt: ReadWrite<u32, Count::Register>,
+        amount: ReadOnly<u32, Amount::Register>,
+        // padding
+        _reserved: [u32; 2],
+    }
+
+    impl EndpointRegisters {
+        pub fn set_buffer<'a>(&'a self, slice: &'a [VolatileCell<u8>]) {
```

Comment:
This type isn't quite right.

When calling `set_buffer`, Rust will be happy to re-borrow each of the arguments with more constrained lifetimes that match despite one (`self`) having a longer lifetime than the other (`slice`). In other words, this type doesn't actually guarantee what we want: that `slice` lives at least as long as `self`.

Here's an example of this going bad: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e775b053628ed63c6613b2645d04dd4d

Instead, the lifetime, `'a` needs to be concrete w.r.t the overall struct. Two ways of doing this:

  1. Change the lifetime to `'static`, which is basically what we always want (since we'll always use these registers to touch hardware that is never "deallocated"

  2. Make `'a` a type-parameter of the struct: `EndpointRegisters<'a>`, which would have to bubble up to the containing struct and likely initialized as `'static` in all practical cases.

Basically think of it as having to put a `&[u8]` as a field in the struct---you'd have to have _some_ lifetime which would either be `'static` or a provided by a type-parameter.

I think (1) is probably best

Reply:
You are totally right about the lifetimes, thanks for checking this! However, using a `'static` lifetime will likely cause issues when crossing the HIL, as we will be able to discuss in the following pull-requests.

It essentially boils down to https://github.com/tock/tock/issues/1074 ("Generic over static lifetimes in HIL interfaces"). Even if all practical cases use `'static` at the end of the day, a generic lifetime is for some reason more flexible in particular when initializing capsules for boards.

So I would go for 2, which shouldn't be problematic.